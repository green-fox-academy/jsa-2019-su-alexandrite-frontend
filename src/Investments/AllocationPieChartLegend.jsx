import React from 'react';
import PropTypes from 'prop-types';
import { Text, View } from 'react-native';
import Row from '../common/Row';
import styles from './styles';

const AllocationPieChartLegend = ({ label, color }) => (
  <Row style={styles.pieChartLegend}>
    <View style={[
      styles.pieChartLegendColorIndicator,
      { backgroundColor: color },
    ]}
    />
    <Text style={styles.pieChartLegendLabel}>{label}</Text>
  </Row>
);

AllocationPieChartLegend.propTypes = {
  label: PropTypes.string.isRequired,
  color: PropTypes.string.isRequired,
};

export default AllocationPieChartLegend;
