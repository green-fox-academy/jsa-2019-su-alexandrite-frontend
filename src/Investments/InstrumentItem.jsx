import React, { useState } from 'react';
import {
  Text,
  Image,
  TouchableHighlight,
} from 'react-native';
import { useNavigation } from 'react-navigation-hooks';
import PropTypes from 'prop-types';
import Card from '../common/Card';
import Column from '../common/Column';
import styles from './styles';
import { moneyAmount2String, decimal2Percentage } from '../common/numbers';
import Row from '../common/Row';
import { DetailsRow } from '../StockDetails/Stats';

const InstrumentItem = ({
  shares, symbol, entryPrice, instrument, index,
}) => {
  const { push } = useNavigation();
  const [descIsExpanded, setDescIsExpanded] = useState(false);
  const unrealizedPL = instrument.marketValue - entryPrice;
  const unrealizedPLPercent = unrealizedPL / entryPrice;
  return (
    <Card
      title={(index === 0) ? 'Instruments' : null}
      style={{ ...styles.instrumentContainer, marginTop: 15 }}
      key={symbol}
    >
      <Row style={styles.companyTitle}>
        <Image
          style={styles.companyLogo}
          source={{ uri: `${instrument.logo}` }}
        />
        <Column style={styles.companyInformation} flex={1}>
          <Text style={styles.companySymbol}>{symbol}</Text>
          <Text style={styles.companyName}>{instrument.company}</Text>
          <TouchableHighlight onPress={() => setDescIsExpanded(!descIsExpanded)} underlayColor="#eee">
            <Text
              style={styles.companyDescription}
              numberOfLines={descIsExpanded ? 0 : 3}
            >
              {instrument.description}
            </Text>
          </TouchableHighlight>
        </Column>
      </Row>
      <Row style={{ marginTop: 15, flexWrap: 'wrap', alignItems: 'center' }}>
        {DetailsRow('Positions', shares)}
        {DetailsRow('Market Value', moneyAmount2String(instrument.marketValue))}
        {DetailsRow('Unrlzd P/L%', `${decimal2Percentage(unrealizedPLPercent)}%`, unrealizedPLPercent > 0 ? '#50B113' : '#FF3636')}
        {DetailsRow('Unrlzd P/L', moneyAmount2String(unrealizedPL), unrealizedPL > 0 ? '#50B113' : '#FF3636')}
      </Row>
      <Row style={{ justifyContent: 'flex-end' }}>
        <TouchableHighlight
          activeOpacity={0.5}
          underlayColor="#0000"
          style={styles.viewMoreButton}
          onPress={() => push('StockDetails', { symbol })}
        >
          <Text style={styles.viewMoreText}>View Details</Text>
        </TouchableHighlight>
      </Row>
    </Card>
  );
};

InstrumentItem.propTypes = {
  shares: PropTypes.number.isRequired,
  symbol: PropTypes.string.isRequired,
  entryPrice: PropTypes.number.isRequired,
  instrument: PropTypes.shape({
    logo: PropTypes.string,
    company: PropTypes.string,
    description: PropTypes.string,
    marketValue: PropTypes.number,
  }).isRequired,
  index: PropTypes.number.isRequired,
};

export default InstrumentItem;
