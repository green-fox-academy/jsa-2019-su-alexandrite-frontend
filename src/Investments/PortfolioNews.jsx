import React from 'react';
import {
  Text,
  ActivityIndicator,
} from 'react-native';
import { useSelector } from 'react-redux';
import styles from './styles';
import NewsCard from '../common/News/NewsCard';

const PortfolioNews = () => {
  const {
    isLoading,
    stocks,
    error,
  } = useSelector((state) => state.investments);

  return (
    !error ? (
      <>
        {isLoading && <ActivityIndicator size="large" style={styles.loading} color="#fff" />}
        {!isLoading && !error && stocks && (
          <NewsCard query={stocks.map(({ symbol }) => symbol)} style={{ marginTop: 15 }} />
        )}
      </>
    ) : <Text>{error.message}</Text>
  );
};

export default PortfolioNews;
