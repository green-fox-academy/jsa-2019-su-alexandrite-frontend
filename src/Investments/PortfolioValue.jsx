import React, { useEffect } from 'react';
import {
  View,
  Text,
  ImageBackground,
  ActivityIndicator,
} from 'react-native';
import { useSelector, useDispatch } from 'react-redux';
import { FontAwesome } from '@expo/vector-icons';
import styles from './styles';
import cardBackground from '../../assets/img/investment/card.png';
import { displayInvestmentsValue } from '../redux/investment/actionCreator';

export default function PortfolioValue() {
  const {
    isLoading,
    totalValue,
    error,
  } = useSelector((state) => state.investments);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(displayInvestmentsValue(1));
  }, []);

  return (
    <ImageBackground
      source={cardBackground}
      style={styles.valueCard}
      imageStyle={{ borderRadius: 6 }}
    >
      {!error ? (
        <>
          {isLoading && <ActivityIndicator size="large" style={styles.loading} color="#fff" />}
          {!isLoading && !error && (
            <>
              <View style={styles.value}>
                <FontAwesome size={22} name="money" color="#fff" />
                <Text style={styles.valueText}>Value</Text>
              </View>
              <Text style={styles.valueNumber}>
                {`$${totalValue}`}
              </Text>
              <Text style={styles.totalText}>Total Investment</Text>
            </>
          )}
        </>
      ) : <Text>{error.message}</Text>}
    </ImageBackground>
  );
}

PortfolioValue.defaultProps = {
  userShares: undefined,
  price: undefined,
};
