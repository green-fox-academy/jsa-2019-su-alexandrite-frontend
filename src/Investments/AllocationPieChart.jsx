import React from 'react';
import { PieChart } from 'react-native-svg-charts';
import PropTypes from 'prop-types';
import Column from '../common/Column';
import AllocationPieChartLegend from './AllocationPieChartLegend';
import styles from './styles';
import Row from '../common/Row';

const AllocationPieChart = ({ data }) => (
  <Row>
    <Column flex={3}>
      <PieChart
        data={data}
        style={styles.pieChart}
        valueAccessor={({ item }) => item.value}
        padAngle={0}
        outerRadius="80%"
        innerRadius="50%"
        sort={(_, b) => (b.key === 'Other' ? 1 : 0)}
      />
    </Column>
    <Column flex={2} style={{ justifyContent: 'space-around' }}>
      {data.map(({ key, svg: { fill: color } }) => (
        <AllocationPieChartLegend key={key} label={key} color={color} />
      ))}
    </Column>
  </Row>
);

AllocationPieChart.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      shares: PropTypes.number,
    }),
  ).isRequired,
};

export default AllocationPieChart;
