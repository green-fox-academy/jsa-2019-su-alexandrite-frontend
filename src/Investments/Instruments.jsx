import React from 'react';
import {
  ActivityIndicator,
} from 'react-native';
import { useSelector } from 'react-redux';
import styles from './styles';
import InstrumentItem from './InstrumentItem';
import ErrorMessage from '../common/ErrorMessage';

const Instruments = () => {
  const {
    isLoading,
    stocks,
    instruments,
    error,
  } = useSelector((state) => state.investments);
  if (error) return <ErrorMessage message={error.message} />;
  if (isLoading) return <ActivityIndicator size="large" style={styles.loading} color="#fff" />;
  return (
    !isLoading && !error && stocks && (
      stocks.map(({ shares, symbol, entryPrice }, i) => (
        <InstrumentItem
          index={i}
          shares={shares}
          symbol={symbol}
          key={symbol}
          entryPrice={entryPrice}
          instrument={instruments[symbol]}
        />
      ))
    )
  );
};

export default Instruments;
