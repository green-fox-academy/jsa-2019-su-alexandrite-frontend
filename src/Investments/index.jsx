import React from 'react';
import { View } from 'react-native';
import commonStyles from '../common/styles';
import SearchButton from '../common/HeaderSearchButton';
import PortfolioValue from './PortfolioValue';

export const navigationOptions = {
  title: 'Investments',
  headerRight: () => <SearchButton />,
};

const Investments = () => (
  <View style={commonStyles.container}>
    <PortfolioValue />
  </View>
);

Investments.navigationOptions = navigationOptions;

export default Investments;
