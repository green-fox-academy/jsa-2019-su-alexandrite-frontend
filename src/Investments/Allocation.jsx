import React from 'react';
import {
  ActivityIndicator,
} from 'react-native';
import { useSelector } from 'react-redux';
import styles from './styles';
import Card from '../common/Card';
import ErrorMessage from '../common/ErrorMessage';
import AllocationPieChart from './AllocationPieChart';

const Allocation = () => {
  const {
    isLoading,
    allocation,
    error,
  } = useSelector((state) => state.investments);
  return (
    <Card title="Allocation" style={{ ...styles.pieChartContainer, marginTop: 15 }}>
      {!error ? (
        <>
          {isLoading && <ActivityIndicator size="large" />}
          {!isLoading && !error && allocation && (
            <AllocationPieChart data={allocation} />
          )}
        </>
      ) : <ErrorMessage message={error.message} />}
    </Card>
  );
};

export default Allocation;
