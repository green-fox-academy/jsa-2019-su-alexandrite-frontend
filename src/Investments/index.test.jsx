import React from 'react';
import renderer from 'react-test-renderer';
import Investments, { navigationOptions } from '.';
import SearchButton from '../common/HeaderSearchButton';

jest.useFakeTimers();

jest.mock('react-redux');
jest.mock('../redux/investment/actionCreator');
jest.mock('./PortfolioValue', () => '<PortfolioValue />');

describe('<Investments />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    const wrapper = renderer.create(
      <Investments />,
    );
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const tree = renderer.create(
      <Investments />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render header search button', () => {
    expect(navigationOptions.headerRight()).toEqual(<SearchButton />);
  });
});
