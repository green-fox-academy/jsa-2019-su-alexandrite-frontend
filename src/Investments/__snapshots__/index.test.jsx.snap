// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Investments /> Should render and match the snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#f4f6f8",
      "flex": 1,
      "padding": 15,
    }
  }
>
  <<PortfolioValue /> />
</View>
`;
